velox/common/memory/SharedArbitrator.cpp

```diff
@@ -347,117 +357,186 @@ bool SharedArbitrator::growCapacity(
+    }
+  });
+  if (freedBytes >= op->targetBytes) {
+    checkedGrow(op->requestRoot, freedBytes, op->targetBytes);
+    freedBytes = 0;
+    return true;
+  }
+  VELOX_CHECK_LT(freedBytes, maxGrowTarget);
+  op->needGlobalArbitration = true;
+  return true;
+}
+
+bool SharedArbitrator::runGlobalArbitration(ArbitrationOperation* op) {
+  VELOX_CHECK(op->needGlobalArbitration);
+  const std::chrono::steady_clock::time_point globalArbitrationStartTime =
+      std::chrono::steady_clock::now();
+  std::lock_guard<std::shared_mutex> l(arbitrationLock_);
+  TestValue::adjust(
+      "facebook::velox::memory::SharedArbitrator::runGlobalArbitration", this);
+  op->waitTimeUs +=
+      std::chrono::duration_cast<std::chrono::microseconds>(
+          std::chrono::steady_clock::now() - globalArbitrationStartTime)
+          .count();
+  checkIfAborted(op->requestRoot);
+
+  if (checkFreeCapacity(op)) {
+    return true;
+  }
+
+  VELOX_CHECK_EQ(op->numGlobalArbitrationAttempts, 0);
+  for (op->numGlobalArbitrationAttempts = 0;;
+       ++op->numGlobalArbitrationAttempts) {
+    if (arbitrateMemory(op)) {
       ++numSucceeded_;
       return true;
     }
-    if (numRetries > 0) {
+    if (op->numGlobalArbitrationAttempts > 0) {
       break;
     }
-    VELOX_CHECK(!requestor->aborted());
-    if (!handleOOM(requestor, targetBytes, candidates)) {
+    VELOX_CHECK(!op->requestRoot->aborted());
+    if (!handleOOM(op)) {
       break;
     }
   }
   VELOX_MEM_LOG(ERROR)
       << "Failed to arbitrate sufficient memory for memory pool "
-      << requestor->name() << ", request " << succinctBytes(targetBytes)
-      << " after " << numRetries
+      << op->requestRoot->name() << ", request "
+      << succinctBytes(op->targetBytes) << " after "
+      << op->numGlobalArbitrationAttempts
       << " retries, Arbitrator state: " << toString();
-  RECORD_METRIC_VALUE(kMetricArbitratorFailuresCount);
-  ++numFailures_;
+  updateArbitrationFailureStats();
   return false;
 }
 
-bool SharedArbitrator::checkCapacityGrowth(
-    const MemoryPool& pool,
-    uint64_t targetBytes) const {
-  return (maxGrowCapacity(pool) >= targetBytes) &&
-      (capacityAfterGrowth(pool, targetBytes) <= capacity_);
+void SharedArbitrator::getGrowTargets(
+    ArbitrationOperation* op,
+    uint64_t& maxGrowTarget,
+    uint64_t& minGrowTarget) {
+  maxGrowTarget = std::min(
+      maxGrowCapacity(*op->requestRoot),
+      std::max(memoryPoolTransferCapacity_, op->targetBytes));
+  minGrowTarget = minGrowCapacity(*op->requestRoot);
 }
 
-bool SharedArbitrator::ensureCapacity(
-    MemoryPool* requestor,
-    uint64_t targetBytes) {
-  if ((targetBytes > capacity_) || (targetBytes > requestor->maxCapacity())) {
+void SharedArbitrator::checkIfAborted(MemoryPool* pool) {
+  if (pool->aborted()) {
+    updateArbitrationFailureStats();
+    VELOX_MEM_POOL_ABORTED("The requestor pool has been aborted");
+  }
+}
+
+bool SharedArbitrator::checkFreeCapacity(ArbitrationOperation* op) {
```

Comment:
checkFreeCapacity looks like a name that only does check without modifying. Here we grow capcity if the requestor itself has enough capacity. Shall we call it maybeGrowFromSelf? or something similar?

Reply:
Updated.

Reply:
This one probably does not need to be member, either